### DISCUSSION TITLE
API Review - some thoughts...

### DISCUSSION THREAD

**Discussion by mbwhite** [hyperledger/fabric-gateway#172] [General]
API minutiae for some of the different sdks

- *Java:* For the async calls would it be worth using CompletableFutures?
- *Node:* all the apis are 'async' nodejs APIs but we have `async submitTx`  and `async submitTxAsync`
   submitTx - resolves when the transaction has been committed
   submitTxAsync - resolves with the commit object with async apis to get the results.. seems a bit confused...
- *All:* `newProposal` only time Proposal is used, to match consistency with submitTransaction/evaluateTranasction  the examples show these being used for transient data... in this context would `newTransaction` or even `newTransactionProposal` ? Understand that there might be more complex  use cases.. but transient would be common
- *All:* why is the `newChaincodeEventsRequest(chaincodeid)` on the network and not the contract object?
- *Alll:* in the docs can we give the equivalent of the `newProposal()` form that submit/evaluateTx are equivalent to?

**Comment by davidkhala** [comment]
> * submitTxAsync - resolves with the commit object with async apis to get the results.. seems a bit confused...

Sorry that I have not yet review source code but upon my first glance of
`submitTxAsync - resolves with the commit object with async apis to get the results.`, does it mean the result of proposal reponse or waiting for transaction id witnessed in block?

**Reply by bestbeforetoday** [reply]
From the current API docs:
> Contract.submit(): Submit a transaction to the ledger and return its result only after it is committed to the ledger.

> Contract.submitAsync(): Submit a transaction to the ledger and return immediately after successfully sending to the orderer. [...] The submitted transaction that is returned can be used to obtain to the transaction result, and to wait for it to be committed to the ledger.

**Comment by bestbeforetoday** [comment]
@mbwhite Great discussion points!

> * _Java:_ For the async calls would it be worth using CompletableFutures?

We wondered about this for the existing Java SDK and again for this one. The trouble with returning `CompletableFuture` (or any `Future`) is that:
1. the SDK needs to manage the treading model and timeouts for execution of these tasks
2. it is harder work for application developers if they just want simple sequential invocations

So right now my feeling is that plain return values are more helpful as they are easier to invoke and give the calling code the flexibility to wrap with whatever asynchronous execution mechanism they prefer, for example:
```java
Proposal proposal = contract.newProposal("txName").addArguments("arg1", "arg2").build();
try {
    Transaction transaction = CompletableFuture.supplyAsync(proposal::endorse).get(30, TimeUnit.SECONDS);
} catch (InterruptedException e) {
   // Handle interrupt
} catch (ExecutionException e) {
    // Handle endorsement failure
} catch (TimeoutException e) {
    // Handle timeout
}
```

> * _Node:_ all the apis are 'async' nodejs APIs but we have `async submitTx`  and `async submitTxAsync`
>   submitTx - resolves when the transaction has been committed
>   submitTxAsync - resolves with the commit object with async apis to get the results.. seems a bit confused...

The idea is that the "async" variants don't block waiting for a commit to be observed on the ledger, and instead return immediately after successfully submitting to the orderer, allowing the client application to do other work with the transaction result while asynchronously waiting for the commit.

There is a strong precedent for function names to end with "Sync" where they are synchronous in the JavaScript sense, rather than taking a callback or returning a Promise, so that seemed like a confusing choice for the `async submit()` method on Contract. To keep the names short and differentiate this method from the one that doesn't block waiting for the commit status, I chose `async submitAsync()` as the name of the Contract method that returns immediately after submit to the orderer, and allows the client to asynchronously wait for the commit. The other languages follow a similar naming pattern for consistency.

I'm keen to hear if anyone has better alternatives for clear and concise naming.

> * _All:_ `newProposal` only time Proposal is used, to match consistency with submitTransaction/evaluateTranasction  the examples show these being used for transient data... in this context would `newTransaction` or even `newTransactionProposal` ? Understand that there might be more complex  use cases.. but transient would be common

Using `newProposal()` you do build a Proposal object/struct, which you can then choose to `evaluate` or `endorse`, as well as obtain the transaction ID prior to either of these actions. Endorsing the Proposal results in a Transaction, which is ready to submit to the orderer, and can provide you with the transaction result. Both the Proposal and Transaction can be explicitly signed using an off-line signer to support cases where the client application cannot perform the signing itself.

`evaluateTransaction` and `submitTransaction` are just convenience methods to allow the proposal flow to be executed in a single call for simple use cases.

> * _All:_ why is the `newChaincodeEventsRequest(chaincodeid)` on the network and not the contract object?

I would actually prefer if it was on the Contract, as it is in the existing SDKs. This does have the potential to cause unexpected behaviour in the existing SDKs, which is why I currently have it on Network in Fabric Gateway.

The problem is that the Contract represents a smart contract, which may be the default (no namespace) smart contract within a given chaincode or one of several (namespaced) smart contracts within a given chaincode. However, there is no distinction between events emitted by different smart contracts. All events emitted by a given chaincode are tagged only with the chaincode name. So if I listen for events emitted by smart contract "A", that listener will also receive events from smart contracts "B" and "C" within that same chaincode. Often people are only using default smart contracts (so there is a 1-to-1 mapping between chaincode and smart contract) and so don't hit this issue, but it's there.

I think the sensible choices are to either:
1. Leave chaincode events on Network and have them behave as they do today; or
2. Add a new event emitting method to the chaincode's Contract API that includes the contract namespace in the event's chaincode name to allow the client to distinguish between events emitted by different smart contracts, and then move chaincode event listening in the Fabric Gateway client API to the Contract.

> * _Alll:_ in the docs can we give the equivalent of the `newProposal()` form that submit/evaluateTx are equivalent to?

That's a good suggestion. I have raised issue #179
